File: 100/swift/42. Trapping Rain Water.swift
class Solution {
  func trap(_ height: [Int]) -> Int {
    var (l, r) = (0, height.count - 1)
    var (lm, rm) = (height[l], height[r])
    var res = 0
    while l<r {
        if lm < rm{
                l += 1
                lm = max(lm, height[l])
                res += lm - height[l]

            }else{
                r -= 1
                rm = max(rm, height[r])
                res += rm - height[r]
            }
    }
    return res
  }
}

File: 100/golang/42. Trapping Rain Water.go
// 42. Trapping Rain Water
// double pointer methods
func trap(height []int) int {
    sum := 0
    for i:=0;i<len(height)-1;i++{
        maxleft:=0
        for j:=i-1;j>=0;j--{
            if(height[j]>maxleft){
                maxleft=height[j]
            }
        }
        maxright:=0
        for j:=i+1;j<len(height);j++{
            if(height[j]>maxright){
                maxright=height[j]
            }
        }
        minH := min(maxleft,maxright)
        if(height[i]<minH){
            sum=sum+(minH-height[i])
        }
        
    }
    return sum
}
